File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.js
/**
 * @param {string} s
 * @param {string[]} words
 * @return {number[]}
 */
var findSubstring = function(s, words) {
    if (!s.length || !words.length) return [];
    const wordLen = words[0].length, totalLen = wordLen * words.length;
    if (s.length < totalLen) return [];
    const wordCount = {};
    for (let w of words) wordCount[w] = (wordCount[w] || 0) + 1;
    const res = [];
    for (let i = 0; i < wordLen; i++) {
        let left = i, count = 0;
        let seen = {};
        for (let j = i; j + wordLen <= s.length; j += wordLen) {
            let w = s.substr(j, wordLen);
            if (wordCount[w]) {
                seen[w] = (seen[w] || 0) + 1;
                count++;
                while (seen[w] > wordCount[w]) {
                    let leftWord = s.substr(left, wordLen);
                    seen[leftWord]--;
                    left += wordLen;
                    count--;
                }
                if (count === words.length) res.push(left);
            } else {
                seen = {};
                count = 0;
                left = j + wordLen;
            }
        }
    }
    return res;
};

File: sol/solution/0001-0100/0030. Substring with Concatenation of All Words/Solution.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words: return []
        word_len, total_len = len(words[0]), len(words) * len(words[0])
        if len(s) < total_len: return []
        word_count = Counter(words)
        res = []
        for i in range(word_len):
            left, count = i, 0
            seen = defaultdict(int)
            for j in range(i, len(s) - word_len + 1, word_len):
                w = s[j:j+word_len]
                if w in word_count:
                    seen[w] += 1
                    count += 1
                    while seen[w] > word_count[w]:
                        left_word = s[left:left+word_len]
                        seen[left_word] -= 1
                        left += word_len
                        count -= 1
                    if count == len(words):
                        res.append(left)
                else:
                    seen.clear()
                    count = 0
                    left = j + word_len
        return res